python/ray/data/_internal/planner/aggregate.py

```diff
@@ -28,20 +29,34 @@ def generate_aggregate_fn(
     """Generate function to aggregate blocks by the specified key column or key
     function.
     """
-    # TODO: validate blocks with AggregateFn._validate.
     if len(aggs) == 0:
         raise ValueError("Aggregate requires at least one aggregation")
 
+    import pyarrow as pa
+
     def fn(
         refs: List[RefBundle],
         ctx: TaskContext,
     ) -> Tuple[List[RefBundle], StatsDict]:
         blocks = []
+        block_schemas: List[Optional[Union[type, pa.lib.Schema]]] = []
+        valid_block_schema: Optional[Union[type, pa.lib.Schema]] = None
         for ref_bundle in refs:
-            for block, _ in ref_bundle.blocks:
+            for block, block_metadata in ref_bundle.blocks:
                 blocks.append(block)
+                if block_metadata.schema is not None and valid_block_schema is None:
+                    valid_block_schema = block_metadata.schema
+                block_schemas.append(block_metadata.schema)
         if len(blocks) == 0:
             return (blocks, {})
+        if isinstance(valid_block_schema, pa.lib.Schema):
+            unified_schema = unify_schemas(block_schemas)
+        else:  # Covers both simple-type and None cases.
+            if isinstance(valid_block_schema, type):
+                assert all([b == valid_block_schema for b in block_schemas])
+            unified_schema = valid_block_schema
```

Comment:
same comment as https://github.com/ray-project/ray/pull/34282/files#r1165901708 .

Reply:
Agreed, let's first merge #34282, since this PR will make use of the same refactored logic. Once that is merged, we can use the same logic here.